### ISSUE TITLE
fastlock_t will always be defined as pthread_mutex_t 

### ISSUE THREAD

**Issue by bturrubiates** [ofiwg/libfabric#798]
fastlock_t is defined in include/fi.h as: 
```
#if defined(PT_LOCK_SPIN)

#define fastlock_t pthread_spinlock_t
#define fastlock_init(lock) pthread_spin_init(lock, PTHREAD_PROCESS_PRIVATE)
#define fastlock_destroy(lock) pthread_spin_destroy(lock)
#define fastlock_acquire(lock) pthread_spin_lock(lock)
#define fastlock_release(lock) pthread_spin_unlock(lock)

#else

#define fastlock_t pthread_mutex_t
#define fastlock_init(lock) pthread_mutex_init(lock, NULL)
#define fastlock_destroy(lock) pthread_mutex_destroy(lock)
#define fastlock_acquire(lock) pthread_mutex_lock(lock)
#define fastlock_release(lock) pthread_mutex_unlock(lock)

#endif /* PT_LOCK_SPIN */
```

As far as I can tell PT_LOCK_SPIN is never defined in any of the configury and is not present in the config.h output. Doing a quick google search only turns up the libfabric use. As a test I ran configure with CFLAGS=-DPT_LOCK_SPIN and got errors. 

@jsquyres @shefty

**Comment by pmmccorm** [comment]
I did this.

What I had in mind was being able to switch between mutex/spinlock with a define. I thought developers could just add "-DPT_LOCK_SPIN" to their CFLAGS and try out their providers with spin locks vs mutex.

**Comment by bturrubiates** [comment]
@pmmccorm 
I see. I had misunderstood the purpose of that statement. It's my understanding that spinlocks will generally perform better for something like an atomic type (I may be wrong). I was experimenting with configure. -DPT_LOCK_SPIN doesn't actually check if it's capable to use those functions so I added some configury in a separate branch: bturrubiates/libfabric@2bb45a6. There's only warnings that need to be fixed. Do we want to keep the --enable-spinlock flag? That way there is an opt-in, and it also checks to ensure the necessary functions exist.

**Comment by pmmccorm** [comment]
The patch looks OK, but I don't think we need to test it: if we have pthreads we can just assume we have pthread spinlocks (and lets cross the pthread-implementation-without-spinlocks bridge when we come to it). But the configure option I still think is a good idea.

Where I was going to go with this:
 Convert more provider code to use fastlock_* types vs bare pthreads.
 Either move away from static mutex inits or figure out a static spinlock init.
 Test, see if there are any measurable perf benefits.

**Comment by bturrubiates** [comment]
Okay, I submitted a PR. Also OS X has pthreads but not pthread spinlocks (which is why I noticed this).

**Comment by pmmccorm** [comment]
@bturrubiates Perfect then, thumbs up from me.

**Comment by pmmccorm** [comment]
One more thing though: if in the future we do something like enable spinlocks by default, will this fall back to sleeping on OS X?

**Comment by bturrubiates** [comment]
Well, we could enable spinlocks by default and try falling back to pthread_mutex if spinlocks aren't available. That's not what my PR currently does. The PR just adds a --enable-spinlock option that will check that the necessary functions exist and error out if they don't.

**Comment by pmmccorm** [comment]
Maybe just deal with that in the future and keep this PR simple..

**Comment by bturrubiates** [comment]
Merged with PR #799
